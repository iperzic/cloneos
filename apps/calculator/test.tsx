import { fireEvent, render } from '@testing-library/react';

import Calculator from './index';

type ButtonPress =
  | '1'
  | '2'
  | '3'
  | '4'
  | '5'
  | '6'
  | '7'
  | '8'
  | '9'
  | '0'
  | '+'
  | '-'
  | '×'
  | '÷'
  | '='
  | 'Clear'
  | 'All Clear'
  | ','
  | '%'
  | '+/-';
type ButtonPresses = ButtonPress[];
type ExpectedDisplay = string;
type TestCase = [ButtonPresses, ExpectedDisplay];

const testCases: TestCase[] = [
  [['1'], '1'],
  [['1', '1'], '11'],
  [['1', '+', '3', '='], '4'],
  [['7', '-', '4', '='], '3'],
  [['5', '×', '6', '='], '30'],
  [['8', '÷', '2', '=', '='], '2'],
  [['1', '8', '÷', '9', '='], '2'],
  [['1', '2', '3'], '123'],
  [['1', '2', '+'], '12'],
  [['1', '2', '3', '+', '4', '='], '127'],
  [['1', '2', '3', '+', '4', '-', '5', '='], '122'],
  [['1', '+', '2', '=', '3', '+', '4', '='], '7'],
  [['1', '2', '3', '+', 'Clear'], '0'],
  [['1', '2', '3', '+', '+'], '123'],
  [['1', '2', '3', '+', '+', '='], '123'],
  [['1', '2', '3', '×', '4', '÷', '5', '='], '98.4'],
  [['9', '×', '9', '=', 'Clear'], '0'],
  [['1', '2', '3', ',', '4', '5', '+', '6', '7', '+/-', '+'], '56.45'],
  [['2', '5', '0', '0', '0', '0', '0', '%'], '25000'],
];
describe('Calculator', () => {
  it('should match snapshot', async () => {
    const { container } = render(<Calculator />);

    expect(container).toMatchSnapshot();
  });

  it.each(testCases)('%#: %s', async (buttonPresses, display) => {
    const { getByRole, getByTestId } = render(<Calculator />);

    buttonPresses.forEach((button) => {
      fireEvent.click(getByRole('button', { name: button }));
    });

    expect(getByTestId('display').textContent).toBe(display);
  });
});
