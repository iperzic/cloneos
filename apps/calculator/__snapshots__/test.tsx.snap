// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator should match snapshot 1`] = `
<div>
  <div
    class="calculator"
  >
    <div
      class="display"
      data-testid="display"
    >
      0
    </div>
    <div
      class="keyboard"
    >
      <button
        aria-label="All Clear"
        class="button buttonUtil"
        type="button"
      >
        AC
      </button>
      <button
        class="button buttonUtil"
        type="button"
      >
        +/-
      </button>
      <button
        class="button buttonUtil"
        type="button"
      >
        %
      </button>
      <button
        class="button buttonOperator"
        type="button"
      >
        รท
      </button>
      <button
        class="button"
        type="button"
      >
        7
      </button>
      <button
        class="button"
        type="button"
      >
        8
      </button>
      <button
        class="button"
        type="button"
      >
        9
      </button>
      <button
        class="button buttonOperator"
        type="button"
      >
        ร
      </button>
      <button
        class="button"
        type="button"
      >
        4
      </button>
      <button
        class="button"
        type="button"
      >
        5
      </button>
      <button
        class="button"
        type="button"
      >
        6
      </button>
      <button
        class="button buttonOperator"
        type="button"
      >
        -
      </button>
      <button
        class="button"
        type="button"
      >
        1
      </button>
      <button
        class="button"
        type="button"
      >
        2
      </button>
      <button
        class="button"
        type="button"
      >
        3
      </button>
      <button
        class="button buttonOperator"
        type="button"
      >
        +
      </button>
      <button
        class="button buttonWide"
        type="button"
      >
        0
      </button>
      <button
        class="button"
        type="button"
      >
        ,
      </button>
      <button
        class="button buttonOperator"
        type="button"
      >
        =
      </button>
    </div>
  </div>
</div>
`;
